diff --git a/OvmfPkg/Library/PlatformDebugLibIoPort/DebugLib.c b/OvmfPkg/Library/PlatformDebugLibIoPort/DebugLib.c
index 4e25f19..4a43869 100644
--- a/OvmfPkg/Library/PlatformDebugLibIoPort/DebugLib.c
+++ b/OvmfPkg/Library/PlatformDebugLibIoPort/DebugLib.c
@@ -16,6 +16,7 @@
 #include <Library/PcdLib.h>
 #include <Library/BaseMemoryLib.h>
 #include <Library/DebugPrintErrorLevelLib.h>
+#include <Library/SerialPortLib.h>
 #include "DebugLibDetect.h"
 
 //
@@ -201,6 +202,7 @@ DebugAssert (
 {
   CHAR8  Buffer[MAX_DEBUG_MESSAGE_LENGTH];
   UINTN  Length;
+  EFI_STATUS                     Status;
 
   //
   // Generate the ASSERT() message in Ascii format
@@ -221,6 +223,11 @@ DebugAssert (
     IoWriteFifo8 (PcdGet16 (PcdDebugIoPort), Length, Buffer);
   }
 
+  // Also send to the serial port
+  Status = SerialPortInitialize ();
+  if (!EFI_ERROR (Status))
+    SerialPortWrite ((UINT8 *)Buffer, Length);
+
   //
   // Generate a Breakpoint, DeadLoop, or NOP based on PCD settings
   //
diff --git a/OvmfPkg/Library/PlatformDebugLibIoPort/PlatformDebugLibIoPort.inf b/OvmfPkg/Library/PlatformDebugLibIoPort/PlatformDebugLibIoPort.inf
index 94ab910..9008815 100644
--- a/OvmfPkg/Library/PlatformDebugLibIoPort/PlatformDebugLibIoPort.inf
+++ b/OvmfPkg/Library/PlatformDebugLibIoPort/PlatformDebugLibIoPort.inf
@@ -34,6 +34,7 @@
   OvmfPkg/OvmfPkg.dec
 
 [LibraryClasses]
+  SerialPortLib
   BaseMemoryLib
   IoLib
   PcdLib
