diff --git a/ArmPlatformPkg/Drivers/NorFlashDxe/NorFlash.c b/ArmPlatformPkg/Drivers/NorFlashDxe/NorFlash.c
index 1b43107..dde0868 100644
--- a/ArmPlatformPkg/Drivers/NorFlashDxe/NorFlash.c
+++ b/ArmPlatformPkg/Drivers/NorFlashDxe/NorFlash.c
@@ -154,8 +154,9 @@ NorFlashEraseSingleBlock (
   }
 
   if (StatusRegister & P30_SR_BIT_ERASE) {
-    DEBUG ((DEBUG_ERROR, "EraseSingleBlock(BlockAddress=0x%08x: Block Erase Error StatusRegister:0x%X\n", BlockAddress, StatusRegister));
-    Status = EFI_DEVICE_ERROR;
+    // Allow a Program Error to succeed to support readonly VARS file
+    // DEBUG((DEBUG_ERROR,"EraseSingleBlock(BlockAddress=0x%08x: Block Erase Error StatusRegister:0x%X\n", BlockAddress, StatusRegister));
+    Status = EFI_SUCCESS;
   }
 
   if (StatusRegister & P30_SR_BIT_BLOCK_LOCKED) {
@@ -210,8 +211,9 @@ NorFlashWriteSingleWord (
   }
 
   if (StatusRegister & P30_SR_BIT_PROGRAM) {
-    DEBUG ((DEBUG_ERROR, "NorFlashWriteSingleWord(WordAddress:0x%X): Program Error\n", WordAddress));
-    Status = EFI_DEVICE_ERROR;
+    // Allow a Program Error to succeed to support readonly VARS file
+    // DEBUG((DEBUG_ERROR,"NorFlashWriteSingleWord(WordAddress:0x%X): Program Error\n",WordAddress));
+    Status = EFI_SUCCESS;
   }
 
   if (StatusRegister & P30_SR_BIT_BLOCK_LOCKED) {
@@ -342,8 +344,9 @@ NorFlashWriteBuffer (
   }
 
   if (StatusRegister & P30_SR_BIT_PROGRAM) {
-    DEBUG ((DEBUG_ERROR, "NorFlashWriteBuffer(TargetAddress:0x%X): Program Error\n", TargetAddress));
-    Status = EFI_DEVICE_ERROR;
+    // Allow a Program Error to succeed to support readonly VARS file
+    // DEBUG((DEBUG_ERROR,"NorFlashWriteBuffer(TargetAddress:0x%X): Program Error\n", TargetAddress));
+    Status = EFI_SUCCESS;
   }
 
   if (StatusRegister & P30_SR_BIT_BLOCK_LOCKED) {
