diff --git a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfig.vfr b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfig.vfr
index e4560c5..c6f8f2a 100644
--- a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfig.vfr
+++ b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfig.vfr
@@ -47,6 +47,7 @@ formset
     //
     // Display of Oneof: 'Secure Boot Mode'
     //
+grayoutif NOT ideqval SECUREBOOT_CONFIGURATION.PhysicalPresent == 1;
     oneof name = SecureBootMode,
           questionid = KEY_SECURE_BOOT_MODE,
           prompt = STRING_TOKEN(STR_SECURE_BOOT_MODE_PROMPT),
@@ -55,6 +56,7 @@ formset
           option text = STRING_TOKEN(STR_STANDARD_MODE),    value = SECURE_BOOT_MODE_STANDARD, flags = DEFAULT;
           option text = STRING_TOKEN(STR_CUSTOM_MODE),      value = SECURE_BOOT_MODE_CUSTOM,   flags = 0;
     endoneof;
+endif;
 
     //
     // Display of 'Current Secure Boot Mode'
diff --git a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf
index 1671d5b..c38c828 100644
--- a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf
+++ b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf
@@ -37,6 +37,7 @@
   MdeModulePkg/MdeModulePkg.dec
   SecurityPkg/SecurityPkg.dec
   CryptoPkg/CryptoPkg.dec
+  OvmfPkg/OvmfPkg.dec
 
 [LibraryClasses]
   BaseLib
@@ -56,6 +57,7 @@
   PeCoffLib
   SecureBootVariableLib
   SecureBootVariableProvisionLib
+  QemuFwCfgSimpleParserLib
 
 [Guids]
   ## SOMETIMES_CONSUMES      ## Variable:L"CustomMode"
diff --git a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c
index 4299a6b..dd07617 100644
--- a/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c
+++ b/SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c
@@ -14,6 +14,7 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/BaseCryptLib.h>
 #include <Library/SecureBootVariableLib.h>
 #include <Library/SecureBootVariableProvisionLib.h>
+#include <Library/QemuFwCfgSimpleParserLib.h>
 
 CHAR16  mSecureBootStorageName[] = L"SECUREBOOT_CONFIGURATION";
 
@@ -3294,6 +3295,23 @@ UpdateSecureBootString (
   return EFI_SUCCESS;
 }
 
+BOOLEAN
+EFIAPI
+IsOciEnv (
+  VOID
+  )
+{
+  EFI_STATUS Status;
+  UINT8      FwCfg_OciEnable = 0;
+
+  // Check for OCI-ENABLE fw-cfg option - which is set only in OCI env
+  Status = QemuFwCfgParseUint8("opt/ovmf/OCI-ENABLE", TRUE, &FwCfg_OciEnable);
+  if (Status == EFI_SUCCESS && FwCfg_OciEnable)
+    return TRUE;
+
+  return FALSE;
+}
+
 /**
   This function extracts configuration from variable.
 
@@ -3337,7 +3355,7 @@ SecureBootExtractConfigFromVariable (
   //
   // If it is Physical Presence User, set the PhysicalPresent to true.
   //
-  if (UserPhysicalPresent ()) {
+  if (UserPhysicalPresent () && !IsOciEnv()) {
     ConfigData->PhysicalPresent = TRUE;
   } else {
     ConfigData->PhysicalPresent = FALSE;
