From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 17 Nov 2015 11:56:06 -0500
Subject: [PATCH ipxe] build: customize configuration

diff --git a/src/config/local/general.h b/src/config/local/general.h
index 5814511..47d0e1a 100644
--- a/src/config/local/general.h
+++ b/src/config/local/general.h
@@ -0,0 +1,5 @@
+#undef ROM_BANNER_TIMEOUT
+#define ROM_BANNER_TIMEOUT     0
+
+/* Enable IPv6. */
+#define NET_PROTO_IPV6
-- 
2.9.3

